Proposals v2 Iteration4 - Discussion system prototype

shamil-gadelshin opened the pull request:
This PR contains a discussion system prototype for the proposals. It's not meant to be merged.

Features:
- implemented in separate substrate module (and crate)
- allows to create and store threads with title
- allows to create and store posts with the text related to an existing thread
- the proposals engine was updated to support a `discussion_thread_id` parameter
- the proposals codex creates a discussion and passes its id to the engine

Major review issues:
- codex, engine, discussion integration. Now, the codex creates a discussion and pass it to the engine
- posts storage. Now, posts are stored in the discussion system in the map (post_id to post) like in the Forum module. Probably we should use another approach (store them inside a thread structure or in double_map(thread_id, post_id to post)

Additional questions:
- Should we create events in the discussion module?
- Is discussion thread mandatory in engine module - `create_proposal()`? Should we add and `ensure_*` call?

shamil-gadelshin committed:
Add discussion module skeleton

shamil-gadelshin committed:
Add create_discussion() API method

- plain implementation (no ensures)
- simple test

shamil-gadelshin committed:
Add a post

shamil-gadelshin committed:
Add virtual workspace for the proposals module

Add virtual workspace Cargo.toml for the proposals module with three modules:
- engine
- codex
- discussion

shamil-gadelshin committed:
Integrate discussion system with codex and engine

shamil-gadelshin committed:
Add comments

bedeho commented on modules/proposals/discussion/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+/// 'Proposal discussion' substrate module Trait
+pub trait Trait: system::Trait {
+    /// Origin from which thread author must come.
+    type ThreadAuthorOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
```
Rename to `ThreadAuthorEnsureOrigin`.

bedeho replied:
Success should not be `AccountId`, it should be an immutable integer identifier, e.g. `ThreadAuthorId`

shamil-gadelshin replied:
I'm not sure such renaming is applicable here: `EnsureOrigin` is a trait same as `Get<>`. Check `Democracy` module trait for example

```
	/// Origin from which any referendum may be cancelled in an emergency.
	type CancellationOrigin: EnsureOrigin<Self::Origin>;

	/// Origin for anyone able to veto proposals.
	type VetoOrigin: EnsureOrigin<Self::Origin, Success=Self::AccountId>;
```

shamil-gadelshin replied:
_Success should not be AccountId, it should be an immutable integer identifier, e.g. ThreadAuthorId_

Answered in general comment below.

bedeho replied:
Great.

bedeho commented on modules/proposals/discussion/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+    type ThreadAuthorOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
+
+    /// Origin from which commenter must come.
+    type PostAuthorOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
```
Rename to `PostAuthorEnsureOrigin`.

shamil-gadelshin replied:
Same as above.

bedeho replied:
Ok.

bedeho commented on modules/proposals/discussion/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    /// Create the discussion
+    pub fn create_discussion(
```
Let's use consistent naming, so `create_thread`.

shamil-gadelshin replied:
Will do

bedeho replied:
Ok.

bedeho commented on modules/proposals/discussion/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+    type PostId: From<u32> + Parameter + Default + Copy;
+
+    /// Type for the thread author id. Should be authenticated by account id.
+    type ThreadAuthorId: From<Self::AccountId> + Parameter + Default;
```
There should not be a `From<AccountId>` here, there need not be any injective mapping from Account space to `ThreadAuthorId`.  E.g. two members can have the same role account at any given time. Currently, this is actually not possible, but we will allow this in the future:

https://github.com/Joystream/substrate-runtime-joystream/issues/93

shamil-gadelshin replied:
Answered in general comment below.

bedeho replied:
Ok.

bedeho commented on modules/proposals/discussion/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+    type ThreadAuthorId: From<Self::AccountId> + Parameter + Default;
+
+    /// Type for the post author id. Should be authenticated by account id.
+    type PostAuthorId: From<Self::AccountId> + Parameter + Default;
```
Same here.

shamil-gadelshin replied:
Answered in general comment below.

bedeho replied:
Ok.

bedeho commented on modules/proposals/codex/src/lib.rs:
```diff
@@ -24,9 +24,10 @@ use rstd::marker::PhantomData;
 
 /// 'Proposals codex' substrate module Trait
-pub trait Trait: system::Trait + proposal_engine::Trait {}
+pub trait Trait: system::Trait + proposal_engine::Trait + proposal_discussion::Trait {}
```
Why is `proposal_discussion::Trait` constraint needed here? I think we do need a trait for the discussion module interface itself, such that calls are mockable, but not sure why module trait is needed.

shamil-gadelshin replied:
I didn't plan to use mocks for the discussion modules. It is a small subsystem designed to be used within proposal modules. And module `Trait` is a Parity default way to use modules.

shamil-gadelshin replied:
Not relevant anymore, because we need to inherit from the `discussion::Trait` because we will store **discussions** to **proposals** map in the **codex** module.

bedeho reviewed (changes requested):
- Perhaps I am misunderstanding, but is the role of the `double_origin` to clone the origin, so that one can be passed on to the discussion module?

- Ids seem to be tied to `u32` for some reason, e.g. as seen by `ThreadCount` and `PostCount`. I think we should either have the ids be fully general or just bite the bullet and alias u32. The former has been the convention thus far. See

https://github.com/Joystream/substrate-runtime-joystream/blob/development/src/membership/members.rs#L17

- Seeing your implementation suggested to me a way we could, in fact, have reused the forum module, but still allowing customized access control. We could do this by making the forum module a so-called instance module, so we could have two separate instances, and one would be connected with the proposal system, and access to use it would be provided by the proposal system through the `*EnsureOrigin` traits. There are still some features which are present in the forum which do not entirely make sense, like the polling, and perhaps even reactions, but by introducing separate origins for each type of action in that module could possibly allow us to only expose the functionality required by the proposal module.

The only problem with this, apart from the fact that there may be unintended costs I am not seeing, is that the new forum module has a lot of unmerged changes that we cannot deploy for Constantinople, hence making independent changes to this forum module will create a merging mess. 

Another option is, if we think we can get the forum to work this way in the future, then we just remove an integrated discussion system as part of the Constantinople release, and just have people use the forum manually until we can deploy a full fix.

**Perhaps we can discuss this last point**

shamil-gadelshin commented:
Yes, `double_origin()` is for `origin` cloning. Our version of the substrate doesn't allow `clone()` for origin (up-to-date version of the substrate fixes it) - I marked this method for deletion when we update our substrate dependency.

shamil-gadelshin commented:
We encountered issues with AccountId and VetoOrigin in the engine module. I suggest the same solution - to postpone it until the Membership and Council modules are separated in their crates.

After that, we can consistently deal with all Origins, AccountIds and other *Id types

shamil-gadelshin commented:
_Ids seem to be tied to u32 for some reason, e.g. as seen by ThreadCount and PostCount. I think we should either have the ids be fully general or just bite the bullet and alias u32. The former has been the convention thus far._

I can probably do ThreadCount and PostCount more general by converting them to the `SimpleArithmetics` trait and adding `Into<u32>` trait. 

But I need `u32` or another exact type to be able to convert T::ThreadId from the discussion module to the T::ThreadId from the engine module. They cannot be absolutely abstract. Another solution would be to inherit the engine Trait from the discussion Trait. This way types would be identical, but I don't want to introduce such dependency. Is there a better option?

shamil-gadelshin committed:
Add update_post() extrinsic

- add update_post() extrinsic with tests
- rename create_discussion() to the create_thread()

shamil-gadelshin commented:
An issue regarding ThreadAuthorId, EnsureOrigin Trait, etc:
https://github.com/Joystream/substrate-runtime-joystream/issues/174

shamil-gadelshin committed:
Change MaxPostEditionNumber from config() to Get<>

- change MaxPostEditionNumber from config() to Get<> parameter in the discussion proposal module

shamil-gadelshin committed:
Migrate posts from map to double map in discussion

shamil-gadelshin committed:
Move discussion dependency from the engine to the codex

shamil-gadelshin committed:
Add tests for thread and post content in discussions

shamil-gadelshin committed:
Add create_thread() title ensure_*()

- add ‘title is empty’ check
- add ‘title is too big’ check

shamil-gadelshin committed:
Add checks for discussion module extrinsics

Add ensure_* for:
- add_post
- update_post

bedeho committed:
Merge pull request #162 from shamil-gadelshin/proposals_v2_iteration3

Proposals v2. Iteration3

shamil-gadelshin committed:
Merge branch 'proposals_v2' into proposals_v2_iteration4

# Conflicts:
#	modules/proposals/codex/src/lib.rs
#	modules/proposals/engine/src/tests/mod.rs

shamil-gadelshin commented:
The code was integrated within the upstream repository
